File: condor/acp/usage.py
"""What a model has read and written, in one shape for every backend (FEAT-120).

Both clients already receive exact usage from their backend — the ACP adapter on
every ``session/prompt`` response and ``usage_update`` notification, pydantic-ai
on every run — and this is the one type they fold it into. It lives in
``condor.acp`` because both clients import it and ``condor.runtime`` already
depends on this package, never the reverse.
"""

from __future__ import annotations

from dataclasses import dataclass, fields
from typing import Any

# Summed by ``+`` and subtracted by ``-``. The context readings are not: they
# are the latest value, and a sum of two occupancies means nothing.
_COUNTERS = (
    "input_tokens",
    "output_tokens",
    "cache_read_tokens",
    "cache_write_tokens",
    "cost_usd",
    "unpriced_turns",
)


@dataclass
class TokenUsage:
    """A running tally of tokens and their price at API rates.

    **``input_tokens`` is inclusive**: every prompt token the model read, cache
    hits and cache writes included, so ``total_tokens`` means "everything the
    model read and wrote" on both backends. pydantic-ai (like OpenAI's
    ``prompt_tokens``) already counts that way. Anthropic does not — its
    ``input_tokens`` excludes both cache counters — so the ACP fold adds them
    back in (:meth:`from_acp`). ``cache_*`` are therefore subsets of
    ``input_tokens``, never additions to it.

    ``cost_usd`` is what the tokens would cost at API prices, summed over
    whatever could be priced. A Claude subscription is not billed per token, so
    this is an estimate and never a spend. ``unpriced_turns`` counts the runs
    whose price could not be known (a local model, an id the bundled price
    table does not have): a chat with both is a lower bound.

    The ACP adapter keeps a background task's result out of the turn's token
    counts and reports only its cost and context. Those tokens are missing here
    while their cost is not; reconstructing them is not worth it.
    """

    input_tokens: int = 0
    output_tokens: int = 0
    cache_read_tokens: int = 0
    cache_write_tokens: int = 0
    cost_usd: float = 0.0
    unpriced_turns: int = 0
    # Latest-value fields: overwritten, never summed.
    context_used: int | None = None
    context_size: int | None = None

    @property
    def total_tokens(self) -> int:
        return self.input_tokens + self.output_tokens

    def is_zero(self) -> bool:
        """Nothing was counted. The context readings do not count as usage."""
        return not any(getattr(self, name) for name in _COUNTERS)

    def __add__(self, other: TokenUsage) -> TokenUsage:
        summed = {
            name: getattr(self, name) + getattr(other, name) for name in _COUNTERS
        }
        return TokenUsage(
            **summed,
            context_used=(
                other.context_used
                if other.context_used is not None
                else self.context_used
            ),
            context_size=(
                other.context_size
                if other.context_size is not None
                else self.context_size
            ),
        )

    def __sub__(self, other: TokenUsage) -> TokenUsage:
        """What ``self`` has that ``other`` had not. Floored at zero.

        The context readings are ``self``'s: a delta taken now carries the
        latest occupancy, not a difference of two.
        """
        return TokenUsage(
            **{
                name: max(getattr(self, name) - getattr(other, name), 0)
                for name in _COUNTERS
            },
            context_used=self.context_used,
            context_size=self.context_size,
        )

    def to_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = {f.name: getattr(self, f.name) for f in fields(self)}
        # Float subtraction leaves dust (0.41 - 0.4 = 0.00999…); a micro-dollar
        # is below anything the readout shows.
        out["cost_usd"] = round(self.cost_usd, 6)
        out["total_tokens"] = self.total_tokens
        return out

    @classmethod
    def from_dict(cls, data: dict | None) -> TokenUsage:
        """Tolerant both ways: unknown keys are ignored, missing ones read as 0.

        A meta written before FEAT-120 has no ``usage`` at all, which reads as
        zero — "unknown before this existed", not a real reading.
        """
        if not isinstance(data, dict):
            return cls()
        usage = cls()
        for name in _COUNTERS:
            value = data.get(name)
            if isinstance(value, (int, float)) and not isinstance(value, bool):
                setattr(usage, name, float(value) if name == "cost_usd" else int(value))
        for name in ("context_used", "context_size"):
            value = data.get(name)
            if isinstance(value, int) and not isinstance(value, bool):
                setattr(usage, name, value)
        return usage

    @classmethod
    def from_acp(cls, usage: Any) -> TokenUsage:
        """A ``session/prompt`` response's ``usage``, with input made inclusive.

        The adapter reports Anthropic's split — ``inputTokens`` excludes
        ``cachedReadTokens`` and ``cachedWriteTokens`` (its own ``totalTokens``
        is the sum of all four) — so they are added back into input here.
        """
        if not isinstance(usage, dict):
            return cls()

        def count(key: str) -> int:
            value = usage.get(key)
            return int(value) if isinstance(value, (int, float)) else 0

        cache_read = count("cachedReadTokens")
        cache_write = count("cachedWriteTokens")
        return cls(
            input_tokens=count("inputTokens") + cache_read + cache_write,
            output_tokens=count("outputTokens"),
            cache_read_tokens=cache_read,
            cache_write_tokens=cache_write,
        )

File: frontend/src/lib/usage.ts
/**
 * A conversation's token total, folded the way the backend folds it (FEAT-120).
 *
 * The dashboard seeds a slot from the conversation's stored total and then adds
 * each turn's `prompt_done.usage` to it, so these rules have to be the Python
 * `TokenUsage.__add__` rules exactly or a live tab would drift from a reloaded
 * one: counters sum, the context reading is the latest one that was reported.
 */

import type { TokenUsage } from "@/lib/api";

const COUNTERS = [
  "input_tokens",
  "output_tokens",
  "cache_read_tokens",
  "cache_write_tokens",
  "cost_usd",
  "unpriced_turns",
] as const;

/** Past this share of the window the context reading turns amber: the moment
 *  `/compact` becomes worth knowing about. */
export const CONTEXT_WARN_RATIO = 0.8;

const num = (v: unknown): number => (typeof v === "number" && Number.isFinite(v) ? v : 0);
const reading = (v: unknown): number | null =>
  typeof v === "number" && Number.isFinite(v) ? v : null;

/**
 * `total` plus one turn. Either side may be partial wire data — a meta written
 * before FEAT-120 carries `{}` — and a missing side changes nothing.
 */
export function addUsage(
  total: Partial<TokenUsage> | null | undefined,
  turn: Partial<TokenUsage> | null | undefined,
): TokenUsage | undefined {
  if (!total && !turn) return undefined;
  const a = total ?? {};
  const b = turn ?? {};
  const next = {} as TokenUsage;
  for (const key of COUNTERS) next[key] = num(a[key]) + num(b[key]);
  next.context_used = reading(b.context_used) ?? reading(a.context_used);
  next.context_size = reading(b.context_size) ?? reading(a.context_size);
  next.total_tokens = next.input_tokens + next.output_tokens;
  return next;
}

/**
 * How the cost figure shows, if at all.
 *
 * - `null` — a chat whose every priced figure is zero *because* nothing could be
 *   priced (a local model). "$0.00" there would be a claim, not a reading.
 * - `lowerBound` — some turns were priced and some were not (a chat that moved
 *   from Claude to a local model), so the figure is at least this much.
 */
export function costDisplay(u: TokenUsage): { lowerBound: boolean } | null {
  if (u.cost_usd <= 0 && u.unpriced_turns > 0) return null;
  return { lowerBound: u.cost_usd > 0 && u.unpriced_turns > 0 };
}

/** `38.2k`, `1.2M` — a token count at a glance. */
export function formatTokens(n: number): string {
  if (n >= 1_000_000) return `${trim(n / 1_000_000)}M`;
  if (n >= 1_000) return `${trim(n / 1_000)}k`;
  return String(Math.round(n));
}

function trim(v: number): string {
  // One decimal below 100, none above: `38.2k`, `200k`, never `200.0k`.
  return v >= 100 ? String(Math.round(v)) : v.toFixed(1).replace(/\.0$/, "");
}
